core/src/main/scala/kafka/coordinator/transaction/TransactionManager.scala

```
@@ -0,0 +1,294 @@
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.locks.ReentrantLock
+
+import kafka.common.Topic
+import kafka.log.LogConfig
+import kafka.server.ReplicaManager
+import kafka.utils.CoreUtils.inLock
+import kafka.utils.{Logging, Pool, Scheduler, ZkUtils}
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.record.{FileRecords, MemoryRecords}
+import org.apache.kafka.common.utils.{Time, Utils}
+
+import scala.collection.mutable
+import scala.collection.JavaConverters._
+
+/*
+ * Transaction manager is part of the transaction coordinator, it manages:
+ *
+ * 1. the transaction log, which is a special internal topic.
+ * 2. the pid metadata including its ongoing transaction status.
+ * 3. the background expiration of the transaction as well as the transactional id to pid mapping.
+ *
+ */
+class TransactionManager(brokerId: Int,
+                         zkUtils: ZkUtils,
+                         scheduler: Scheduler,
+                         replicaManager: ReplicaManager,
+                         config: TransactionConfig,
+                         time: Time) extends Logging {
+
+  this.logIdent = "[Transaction Log Manager " + brokerId + "]: "
+
+  /* number of partitions for the transaction log topic */
+  private val transactionTopicPartitionCount = getTransactionTopicPartitionCount
+
+  /* lock protecting access to loading and owned partition sets */
+  private val partitionLock = new ReentrantLock()
+
+  /* partitions of transaction topic that are assigned to this manager, partition lock should be called BEFORE accessing this set */
+  private val ownedPartitions: mutable.Set[Int] = mutable.Set()
+
+  /* partitions of transaction topic that are being loaded, partition lock should be called BEFORE accessing this set */
+  private val loadingPartitions: mutable.Set[Int] = mutable.Set()
+
+  /* partitions of transaction topic that are corrupted; should always be an empty set under normal operations */
```

I think we discussed offline getting rid of this and simply failing the TC. The problem is that we don't know which partition actually has the problem without some more complicated logic. Since we don't expect this to happen in practice (except in the case of bugs), I think failing fast is reasonable instead of limping along in a weird state.